JDK-8318119: Invalid narrow Klass base on aarch64 post 8312018

tstuefe opened the pull request:
When reserving space for class space or the combined CDS+class space, provide a fallback on aarch64 when all prior attempts to reserve space failed. Now the logic goes like this:

1) (only if not CDS) attempt to reserve in low-address regions
2) attempt to reserve at a 32G-aligned address at randomized attach points within the whole of the available address space
3) (new) attempt to reserve 4G-aligned by using os::reserve_aligned.

(3) is similar to (2), but (2) may fail if we are very unlucky and the address space is very populated. The difference between (2) and (3) is that with (3), the kernel chooses the placement of the VMA, and we rely on ASLR to provide entropy.

The patch also fixes os::reserve_memory_aligned() to remove the assertion that *size* needs to be aligned. There is no need for that, and for large alignments this will blow up the required over-allocation size even further.

The patch also fixes up CompressedKlassPointers::is_valid_base(). The whole purpose of this function is to filter out possibly invalid base addresses when someone manually specifies SharedBaseAddress. The problem with this function is that the encoding base alone is not sufficient to determine its validity, since for that one needs to know the shift value too, and for that the boundaries of the future encoding range. I changed this function to give the best answer possible at that point, see comment.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Integration blocker
&nbsp;⚠️ Title mismatch between PR and JBS for issue [JDK-8318119](https://bugs.openjdk.org/browse/JDK-8318119)

### Issue
 * [JDK-8318119](https://bugs.openjdk.org/browse/JDK-8318119): Invalid narrow Klass base on aarch64 post JDK-8312018 (**Bug** - P2) ⚠️ Title mismatch between PR and JBS.


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16215/head:pull/16215` \
`$ git checkout pull/16215`

Update a local copy of the PR: \
`$ git checkout pull/16215` \
`$ git pull https://git.openjdk.org/jdk.git pull/16215/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16215`

View PR using the GUI difftool: \
`$ git pr show -t 16215`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16215.diff">https://git.openjdk.org/jdk/pull/16215.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16215#issuecomment-1773667992)

tstuefe commented on src/hotspot/share/memory/metaspace.cpp:
```diff
@@ -627,11 +627,28 @@ ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t siz
+#ifdef AARCH64
+    if (result == nullptr) {
+      // If that failed, attempt to allocate at any properly aligned address. Try first an alignment that
+      // includes shift (e.g. 32GB). Failing that, try an alignment that does not include the shift (4GB).
```
The way movk mode is currently implemented is to apply the right-shifted base to the narrow Klass ID, then to left-shift the thing to get Klass*

https://github.com/openjdk/jdk/blob/d0ea2a51111bd5de5a6465e7de6a4950aae89c71/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L4723-L4731

which requires the *right-shifted* base to be movk-able. So it must have the lower 32+shift bits zero:

https://github.com/openjdk/jdk/blob/d0ea2a51111bd5de5a6465e7de6a4950aae89c71/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L4656-L4659

hence the 32g alignment.

```
movw dst, src
movk dst, right-shifted-base to q3
shl dst, dst, 3
```

Note that this patch is still draft. I wanted to add another movk decoding variant, would work with a 4GB aligned base, just to increase our options. 

It would apply the unshifted base to the left-shifted narrow Klass ID:

```
shl dst, src, 3
movk dst, unshifted-base
```

and that would even need one less operation. 

Only problem is that it limits the value range of the narrow Klass ID, since it must not spill over into the third quadrant, so its upper (shift) bits must be zero. For shift=3, it means narrowKlass must be <= 2^(32-3). I think this is fine though, since we still can address 4 GB of class space. In other words, we should have never seen such high narrow Klass IDs anyway.

theRealAph replied:
> The way movk mode is currently implemented is to apply the right-shifted base to the narrow Klass ID, then to left-shift the thing to get Klass*

There's no need to left shift the Klass if we're using `movk` mode. Shift should be zero. We should always use a zero shift, 4G aligned, class space. That's how it was intended to work when I wrote it.

tstuefe commented:
Anyone?

theRealAph commented:
> Anyone?

I find those comments utterly incomprehensible.

theRealAph commented on src/hotspot/share/oops/compressedKlass.cpp:
```diff
@@ -90,19 +90,19 @@ void CompressedKlassPointers::initialize(address addr, size_t len) {
+  //  to the base it is not possible to say more. If the base is not 0, we assume
+  //  the range starts at the specified encoding base and its size will not be
+  //  larger than 4GB).
+  return is_aligned(p, 4 * G);
```
Does this mean that the class metadata might be late than 4 Gbyte?

tstuefe replied:
Sorry, I don't understand this question.

theRealAph replied:
Argh, typo. Does this mean that the class metadata might be more than 4G in size? Is tha the risk?

tstuefe replied:
Currently not. Class space is limited to 3Gb, CDS to 1Gb, they are placed adjacent to each other to not surpass 4G.

theRealAph replied:
So what does the word "probably" refer to?

tstuefe replied:
This function attempts to answer "is this address a valid encoding base". Strictly speaking, that question cannot be answered in isolation. The correct question "given a specific Klass range, can a valid encoding scheme be found with this base and any valid shift?".

"Probably" in this context means "This base would work, assuming a future Klass range that starts at the given encoding base and its length does not surpass the maximum encoding range".

This function is used just for one valid use case: verifying the user input for -XX:SharedBaseAddress. That happens at a point where we have not allocated anything yet, so we don't know the future encoding range.

I am really considering removing this function. It is difficult to explain, of not much use, and also slightly wrong since it ignores the base==valid EOR immediate case.

Side note about non-zero immediate + non-zero shift:

The Klass range does not have to start at the encoding base. For encoding base == zero it obviously does not. But also for non-zero encoding bases this could in theory make sense, at least on aarch64 that has no fallback add-base mode.

For example, given a Klass range starting just below 8 GB due to ASLR, this normally would lead to VM exit since such an address is unusable with MOVK, unless it fits EOR mode. But we could use 4GB as encoding base, and a non-zero shift that enlarges the encoding range, and that would still work.

We have not done this so far, and there are things to consider (e.g. it would make sense to reserve and protect a single page at encoding base, to get NULL-reference crashes), but it is possible. That is why I am hesitant to hard-code 4GB as encoding range limit.

theRealAph replied:
> This function is used just for one valid use case: verifying the
> user input for -XX:SharedBaseAddress. That happens at a point where
> we have not allocated anything yet, so we don't know the future
> encoding range.

So this function is simply asking a question that cannot be answered
without actually trying it.

> I am really considering removing this function. It is difficult to
> explain, of not much use, and also slightly wrong since it ignores
> the base==valid EOR immediate case.

Please do! If there is to be an answer about whether a base is valid,
let it happen just after a failed attempted allocation.

> Side note about non-zero immediate + non-zero shift:
> 
> The Klass range does not have to start at the encoding base. For
> encoding base == zero it obviously does not. But also for non-zero
> encoding bases this could in theory make sense, at least on aarch64
> that has no fallback add-base mode.
> 
> For example, given a Klass range starting just below 8 GB due to
> ASLR, this normally would lead to VM exit since such an address is
> unusable with MOVK, unless it fits EOR mode.

Surely we'd scan upwards until a successful allocation. Why does it
matter where the Klass range starts? Or is the problem that we're
trying hard to honour a -XX: user request, for some reason? What would
be the purpose?

theRealAph replied:
Oh, wow. I never heard about "Oracle does not want us to do that for security reasons. We are to provide some form of ASLR." Sorry, this is all news to me.

theRealAph commented on src/hotspot/share/memory/metaspace.cpp:
```diff
@@ -627,22 +627,35 @@ ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t siz
     result = os::attempt_reserve_memory_between((char*)min, (char*)max, size, alignment, randomize);
+
+#ifdef AARCH64
+    if (result == nullptr) {
```
Why not do the optimized (for AArch64) stuff before trying the immediate + shift?

theRealAph replied:
> Why not do the optimized (for AArch64) stuff before trying the immediate + shift?

For AArch64, the optimum is zero base, no shift, then immediate base, no shift, then shift, and last of all immediate + shift.

theRealAph replied:
> > > Why not do the optimized (for AArch64) stuff before trying the immediate + shift?
> > 
> > For AArch64, the optimum is zero base, no shift, then immediate base, no shift, then shift, and last of all immediate + shift.
> 
> Wait, immediate + shift=0 before no immediate=0 + shift? Why? Looking at https://dougallj.github.io/applecpu/firestorm-int.html , both movk and lsl need the same number of cycles. Why would movk be better than lsl?

I agree, good point. Those are equal on modern AArch64 CPUs. LSL (immediate) is an alias for UBFM, and I had a vague memory that was slower, but I don't think that's so now, if ever it was.

> The reason I ask is that this would mean switching up the order of things I do in Metaspace::reserve_address_space_for_compressed_classes. That function attempts to do the best for all platforms, and all other platforms prefer zero immediate + shift. It would mean even more ifdefs than we have before.

I see.

> (the best way would be to move this function into CPU specific files. We would have some duplication, but at least every CPU can then decide how to layout memory, and the comments would not have to be so darn vague. 

Ah, it's not just me, then.

I think you at least have to say what you think might fail, rather than we can say no more. Something like "If the address space is already highly allocated we might fail." 

> But last time I attempted that I ran into objections from reviewers).

Shame. This is very CPU specific.

tstuefe replied:
>> (the best way would be to move this function into CPU specific files. We would have some duplication, but at least every CPU can then decide how to layout memory, and the comments would not have to be so darn vague.

> Ah, it's not just me, then.

No, I really dislike this too. Forcing something that is inherently CPU specific in a CPU agnostic form, and then having to explain it every time.

> Shame. This is very CPU specific.

I may retry making this CPU specific, once I have some cycles left.

tstuefe commented:
> > Anyone?
> 
> I find those comments utterly incomprehensible.

Which ones exactly?

tstuefe commented:
@theRealAph I remove the questionable `CompressedKlassPointers::is_valid_base()` which in hindsight confused more than it helped. I removed all asserts using that function, since on the one platform where it matters we will run against other asserts early if encoding does not work.

For the one legitimate use of that function, validating -XX:SharedBaseAddress, I just added an aarch64-specific line and a clear comment. I feel this is much clearer to the code reader. Sometimes ifdef is the best way to go.

I also added some comments to the reservation function. Please check if this agrees with you.

tstuefe commented:
I am withdrawing this in favour of a clearer, broader solution.
